server/src/main/java/io/squashql/spring/web/rest/QueryController.java

```diff
@@ -47,51 +45,64 @@ public QueryController(QueryEngine<?> queryEngine, Optional<Supplier<SquashQLUse
-    return ResponseEntity.ok(new PivotTableQueryResultDto(cells, pt.rows, pt.columns, pt.values, pt.hiddenTotals));
+    return formatResponseEntity(pt, contentType, () -> new PivotTableQueryResultDto(cells, pt.rows, pt.columns, pt.values, pt.hiddenTotals));
   }
 
   @PostMapping(MAPPING_QUERY_MERGE)
-  public ResponseEntity<QueryResultDto> executeAndMerge(@RequestBody QueryMergeDto queryMergeDto) {
+  public ResponseEntity<?> executeAndMerge(@RequestBody QueryMergeDto queryMergeDto, @RequestHeader("Accept") String contentType) {
     Table table = this.queryExecutor.executeQueryMerge(
             queryMergeDto,
             this.squashQLUserSupplier == null ? null : this.squashQLUserSupplier.get());
-    return ResponseEntity.ok(createQueryResultDto(table, CacheStatsDto.builder(), queryMergeDto.minify));
+    return formatResponseEntity(table, contentType, () -> createQueryResultDto(table, CacheStatsDto.builder(), queryMergeDto.minify));
   }
 
   @PostMapping(MAPPING_QUERY_MERGE_PIVOT)
-  public ResponseEntity<PivotTableQueryResultDto> executeQueryMergePivot(@RequestBody PivotTableQueryMergeDto pivotTableQueryMergeDto) {
+  public ResponseEntity<?> executeQueryMergePivot(@RequestBody PivotTableQueryMergeDto pivotTableQueryMergeDto, @RequestHeader("Accept") String contentType) {
     PivotTable pt = this.queryExecutor.executePivotQueryMerge(
             pivotTableQueryMergeDto,
             this.squashQLUserSupplier == null ? null : this.squashQLUserSupplier.get()
     );
     List<Map<String, Object>> cells = PivotTableUtils.generateCells(pt, pivotTableQueryMergeDto.query.minify);
-    return ResponseEntity.ok(new PivotTableQueryResultDto(cells, pt.rows, pt.columns, pt.values, pt.hiddenTotals));
+    return formatResponseEntity(pt, contentType, () -> new PivotTableQueryResultDto(cells, pt.rows, pt.columns, pt.values, pt.hiddenTotals));
   }
 
   @PostMapping(MAPPING_QUERY_JOIN_EXPERIMENTAL)
-  public ResponseEntity<QueryResultDto> executeQueryJoin(@RequestBody QueryJoinDto queryJoinDto) {
+  public ResponseEntity<?> executeQueryJoin(@RequestBody QueryJoinDto queryJoinDto, @RequestHeader("Accept") String contentType) {
     Table table = this.queryExecutor.executeExperimentalQueryMerge(queryJoinDto);
-    return ResponseEntity.ok(createQueryResultDto(table, CacheStatsDto.builder(), queryJoinDto.minify));
+    return formatResponseEntity(table, contentType, () -> createQueryResultDto(table, CacheStatsDto.builder(), queryJoinDto.minify));
+  }
+
+  public ResponseEntity<?> formatResponseEntity(Renderable table, String contentType, Supplier<?> toJson) {
+    switch (contentType) {
+      case "application/json": return ResponseEntity.ok(toJson.get());
+      case "text/csv": {
+        return ResponseEntity.ok()
+          .header("Content-Type", contentType)
+          .header("Content-Disposition", "attachment; filename=\"result.csv\"")
+          .body(table.toCSV());
```

Comment:
Just a remark, you can keep it like this for now: you could hit the maximum size of a String (I've seen it in the past with big export) and create a lot of transient memory. 
Usually, the table content is streamed into a file stored in a temporary format that can be downloaded by the user.